diff --git a/configure b/configure
index fc11f95..2207807 100755
--- a/configure
+++ b/configure
@@ -364,6 +364,8 @@ External library support:
   --disable-nvdec          disable Nvidia video decoding acceleration (via hwaccel) [autodetect]
   --disable-nvenc          disable Nvidia video encoding code [autodetect]
   --enable-rkmpp           enable Rockchip Media Process Platform code [no]
+  --enable-stcodec         enable Spacemit Media Process Platform code [no]
+  --enable-stcodec         enable Spacemit Media Process Platform code [no]
   --disable-v4l2-m2m       disable V4L2 mem2mem code [autodetect]
   --disable-v4l2-request   disable V4L2 Request API code [autodetect]
   --disable-vaapi          disable Video Acceleration API (mainly Unix/Intel) code [autodetect]
@@ -2059,6 +2061,8 @@ EXTERNAL_LIBRARY_VERSION3_LIST="
     libvo_amrwbenc
     mbedtls
     rkmpp
+    stcodec
+    stcodec
 "
 
 EXTERNAL_LIBRARY_GPLV3_LIST="
@@ -3665,6 +3669,8 @@ h264_oh_encoder_deps="ohcodec"
 h264_qsv_decoder_select="h264_mp4toannexb_bsf qsvdec"
 h264_qsv_encoder_select="atsc_a53 qsvenc"
 h264_rkmpp_decoder_deps="rkmpp"
+h264_stcodec_decoder_deps="stcodec"
+h264_stcodec_decoder_deps="stcodec"
 h264_rkmpp_decoder_select="h264_mp4toannexb_bsf"
 h264_rkmpp_encoder_deps="rkmpp"
 h264_vaapi_encoder_select="atsc_a53 cbs_h264 vaapi_encode"
@@ -3690,6 +3696,8 @@ hevc_oh_encoder_deps="ohcodec"
 hevc_qsv_decoder_select="hevc_mp4toannexb_bsf qsvdec"
 hevc_qsv_encoder_select="hevcparse qsvenc"
 hevc_rkmpp_decoder_deps="rkmpp"
+hevc_stcodec_decoder_deps="stcodec"
+hevc_stcodec_decoder_deps="stcodec"
 hevc_rkmpp_decoder_select="hevc_mp4toannexb_bsf"
 hevc_rkmpp_encoder_deps="rkmpp"
 hevc_vaapi_encoder_deps="VAEncPictureParameterBufferHEVC"
@@ -3732,6 +3740,10 @@ vp8_mediacodec_decoder_deps="mediacodec"
 vp8_mediacodec_encoder_deps="mediacodec"
 vp8_qsv_decoder_select="qsvdec"
 vp8_rkmpp_decoder_deps="rkmpp"
+vp8_stcodec_decoder_deps="stcodec"
+vp9_stcodec_decoder_deps="stcodec"
+vp8_stcodec_decoder_deps="stcodec"
+vp9_stcodec_decoder_deps="stcodec"
 vp8_vaapi_encoder_deps="VAEncPictureParameterBufferVP8"
 vp8_vaapi_encoder_select="vaapi_encode"
 vp8_v4l2m2m_decoder_deps="v4l2_m2m vp8_v4l2_m2m"
@@ -7584,6 +7596,8 @@ enabled rkmpp             && { require_pkg_config rkmpp "rockchip_mpp >= 1.3.8"
                                  die "ERROR: rkmpp requires --enable-libdrm"; }
                              }
 enabled vapoursynth       && require_headers "vapoursynth/VSScript4.h vapoursynth/VapourSynth4.h"
+enabled stcodec
+enabled stcodec
 
 if enabled gcrypt; then
     GCRYPT_CONFIG="${cross_prefix}libgcrypt-config"
diff --git a/libavcodec/Makefile b/libavcodec/Makefile
index d3ffcc0..9dbe16a 100644
--- a/libavcodec/Makefile
+++ b/libavcodec/Makefile
@@ -446,6 +446,8 @@ OBJS-$(CONFIG_H264_OH_ENCODER)         += ohcodec.o ohenc.o
 OBJS-$(CONFIG_H264_QSV_DECODER)        += qsvdec.o
 OBJS-$(CONFIG_H264_QSV_ENCODER)        += qsvenc_h264.o
 OBJS-$(CONFIG_H264_RKMPP_DECODER)      += rkmppdec.o
+OBJS-$(CONFIG_H264_STCODEC_DECODER)    += stcodecdec.o
+OBJS-$(CONFIG_H264_STCODEC_DECODER)    += stcodecdec.o
 OBJS-$(CONFIG_H264_RKMPP_ENCODER)      += rkmppenc.o
 OBJS-$(CONFIG_H264_VAAPI_ENCODER)      += vaapi_encode_h264.o h264_levels.o \
                                           h2645data.o hw_base_encode_h264.o
@@ -476,6 +478,8 @@ OBJS-$(CONFIG_HEVC_OH_ENCODER)         += ohcodec.o ohenc.o
 OBJS-$(CONFIG_HEVC_QSV_DECODER)        += qsvdec.o
 OBJS-$(CONFIG_HEVC_QSV_ENCODER)        += qsvenc_hevc.o hevc/ps_enc.o
 OBJS-$(CONFIG_HEVC_RKMPP_DECODER)      += rkmppdec.o
+OBJS-$(CONFIG_HEVC_STCODEC_DECODER)    += stcodecdec.o
+OBJS-$(CONFIG_HEVC_STCODEC_DECODER)    += stcodecdec.o
 OBJS-$(CONFIG_HEVC_RKMPP_ENCODER)      += rkmppenc.o
 OBJS-$(CONFIG_HEVC_VAAPI_ENCODER)      += vaapi_encode_h265.o h265_profile_level.o \
                                           h2645data.o hw_base_encode_h265.o
@@ -815,6 +819,10 @@ OBJS-$(CONFIG_VP8_MEDIACODEC_DECODER)  += mediacodecdec.o
 OBJS-$(CONFIG_VP8_MEDIACODEC_ENCODER)  += mediacodecenc.o
 OBJS-$(CONFIG_VP8_QSV_DECODER)         += qsvdec.o
 OBJS-$(CONFIG_VP8_RKMPP_DECODER)       += rkmppdec.o
+OBJS-$(CONFIG_VP8_STCODEC_DECODER)     += stcodecdec.o
+OBJS-$(CONFIG_VP9_STCODEC_DECODER)     += stcodecdec.o
+OBJS-$(CONFIG_VP8_STCODEC_DECODER)     += stcodecdec.o
+OBJS-$(CONFIG_VP9_STCODEC_DECODER)     += stcodecdec.o
 OBJS-$(CONFIG_VP8_VAAPI_ENCODER)       += vaapi_encode_vp8.o
 OBJS-$(CONFIG_VP8_V4L2M2M_DECODER)     += v4l2_m2m_dec.o
 OBJS-$(CONFIG_VP8_V4L2M2M_ENCODER)     += v4l2_m2m_enc.o
diff --git a/libavcodec/allcodecs.c b/libavcodec/allcodecs.c
index 47eded2..0bc02c5 100644
--- a/libavcodec/allcodecs.c
+++ b/libavcodec/allcodecs.c
@@ -154,12 +154,16 @@ extern const FFCodec ff_h264_mediacodec_encoder;
 extern const FFCodec ff_h264_mmal_decoder;
 extern const FFCodec ff_h264_qsv_decoder;
 extern const FFCodec ff_h264_rkmpp_decoder;
+extern const FFCodec ff_h264_stcodec_decoder;
+extern const FFCodec ff_h264_stcodec_decoder;
 extern const FFCodec ff_h264_rkmpp_encoder;
 extern const FFCodec ff_hap_encoder;
 extern const FFCodec ff_hap_decoder;
 extern const FFCodec ff_hevc_decoder;
 extern const FFCodec ff_hevc_qsv_decoder;
 extern const FFCodec ff_hevc_rkmpp_decoder;
+extern const FFCodec ff_hevc_stcodec_decoder;
+extern const FFCodec ff_hevc_stcodec_decoder;
 extern const FFCodec ff_hevc_rkmpp_encoder;
 extern const FFCodec ff_hevc_v4l2m2m_decoder;
 extern const FFCodec ff_hnm4_video_decoder;
@@ -375,9 +379,13 @@ extern const FFCodec ff_vp6f_decoder;
 extern const FFCodec ff_vp7_decoder;
 extern const FFCodec ff_vp8_decoder;
 extern const FFCodec ff_vp8_rkmpp_decoder;
+extern const FFCodec ff_vp8_stcodec_decoder;
+extern const FFCodec ff_vp8_stcodec_decoder;
 extern const FFCodec ff_vp8_v4l2m2m_decoder;
 extern const FFCodec ff_vp9_decoder;
 extern const FFCodec ff_vp9_rkmpp_decoder;
+extern const FFCodec ff_vp9_stcodec_decoder;
+extern const FFCodec ff_vp9_stcodec_decoder;
 extern const FFCodec ff_vp9_v4l2m2m_decoder;
 extern const FFCodec ff_vqa_decoder;
 extern const FFCodec ff_vqc_decoder;
diff --git a/libavcodec/stcodecdec.c b/libavcodec/stcodecdec.c
new file mode 100644
index 0000000..1265fe6
--- /dev/null
+++ b/libavcodec/stcodecdec.c
@@ -0,0 +1,787 @@
+/*
+ * Spacemit MPP Video Decoder
+ * Copyright 2022-2023 SPACEMIT. All rights reserved.
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <drm_fourcc.h>
+#include <pthread.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "avcodec.h"
+#include "codec_internal.h"
+#include "decode.h"
+#include "hwconfig.h"
+#include "libavutil/buffer.h"
+#include "libavutil/common.h"
+#include "libavutil/frame.h"
+#include "libavutil/hwcontext.h"
+#include "libavutil/hwcontext_drm.h"
+#include "libavutil/imgutils.h"
+#include "libavutil/log.h"
+#include "libavutil/mem.h"
+#include "vdec.h"
+
+#define FREESLOTS_THRESHOLD 4
+#define DROPFRAME_NUM_AFTER_FLUSH 1
+
+typedef struct {
+  MppVdecCtx* pVdecCtx;
+  MppVdecPara* pVdecPara;
+  MppPacket* pPacket;
+  MppFrame* pFrame;
+  enum AVPixelFormat ePixFmt;
+
+  char eos_reached;
+  char no_pts;
+  char first_packet;
+  int64_t duration;
+  int64_t use_dts;
+  char flushed;
+
+  AVBufferRef* frames_ref;
+  AVBufferRef* device_ref;
+} STCODECDecoder;
+
+typedef struct {
+  AVClass* av_class;
+  AVBufferRef* decoder_ref;
+} STCODECDecodeContext;
+
+typedef struct {
+  MppFrame* pFrame;
+  AVBufferRef* decoder_ref;
+} STCODECFrameContext;
+
+static MppPixelFormat stcodec_get_pixelformat(AVCodecContext* avctx) {
+  switch (avctx->pix_fmt) {
+    case AV_PIX_FMT_NV12:
+      return PIXEL_FORMAT_NV12;
+    case AV_PIX_FMT_YUV420P:
+      return PIXEL_FORMAT_I420;
+    case AV_PIX_FMT_DRM_PRIME:
+      return PIXEL_FORMAT_NV12;
+    default:
+      return PIXEL_FORMAT_UNKNOWN;
+  }
+}
+
+static MppCodingType stcodec_get_codingtype(AVCodecContext* avctx) {
+  switch (avctx->codec_id) {
+    case AV_CODEC_ID_H264:
+      return CODING_H264;
+    case AV_CODEC_ID_HEVC:
+      return CODING_H265;
+    case AV_CODEC_ID_VP8:
+      return CODING_VP8;
+    case AV_CODEC_ID_VP9:
+      return CODING_VP9;
+    case AV_CODEC_ID_MJPEG:
+      return CODING_MJPEG;
+    case AV_CODEC_ID_VC1:
+      return CODING_VC1;
+    case AV_CODEC_ID_MPEG2VIDEO:
+      return CODING_MPEG2;
+    case AV_CODEC_ID_MPEG4:
+      return CODING_MPEG4;
+    case AV_CODEC_ID_AVS:
+      return CODING_AVS;
+    case AV_CODEC_ID_AVS2:
+      return CODING_AVS2;
+    default:
+      return CODING_UNKNOWN;
+  }
+}
+
+static MppProfileType stcodec_get_profiletype(AVCodecContext* avctx) {
+  av_log(avctx, AV_LOG_DEBUG, "profile = %d\n", avctx->profile);
+  if (avctx->codec_id == AV_CODEC_ID_MPEG2VIDEO) {
+    switch (avctx->profile) {
+      case AV_PROFILE_MPEG2_422:
+        return PROFILE_MPEG2_422;
+      case AV_PROFILE_MPEG2_HIGH:
+        return PROFILE_MPEG2_HIGH;
+      case AV_PROFILE_MPEG2_SNR_SCALABLE:
+        return PROFILE_MPEG2_SNR_SCALABLE;
+      case AV_PROFILE_MPEG2_MAIN:
+        return PROFILE_MPEG2_MAIN;
+      case AV_PROFILE_MPEG2_SIMPLE:
+        return PROFILE_MPEG2_SIMPLE;
+      default:
+        return PROFILE_UNKNOWN;
+    }
+  } else if (avctx->codec_id == AV_CODEC_ID_VC1) {
+    switch (avctx->profile) {
+      case AV_PROFILE_VC1_SIMPLE:
+        return PROFILE_VC1_SIMPLE;
+      case AV_PROFILE_VC1_MAIN:
+        return PROFILE_VC1_MAIN;
+      case AV_PROFILE_VC1_COMPLEX:
+        return PROFILE_VC1_COMPLEX;
+      case AV_PROFILE_VC1_ADVANCED:
+        return PROFILE_VC1_ADVANCED;
+      default:
+        return PROFILE_UNKNOWN;
+    }
+  } else if (avctx->codec_id == AV_CODEC_ID_MJPEG) {
+    switch (avctx->profile) {
+      case AV_PROFILE_MJPEG_HUFFMAN_BASELINE_DCT:
+        return PROFILE_MJPEG_HUFFMAN_BASELINE_DCT;
+      case AV_PROFILE_MJPEG_HUFFMAN_EXTENDED_SEQUENTIAL_DCT:
+        return PROFILE_MJPEG_HUFFMAN_EXTENDED_SEQUENTIAL_DCT;
+      case AV_PROFILE_MJPEG_HUFFMAN_PROGRESSIVE_DCT:
+        return PROFILE_MJPEG_HUFFMAN_PROGRESSIVE_DCT;
+      case AV_PROFILE_MJPEG_HUFFMAN_LOSSLESS:
+        return PROFILE_MJPEG_HUFFMAN_LOSSLESS;
+      case AV_PROFILE_MJPEG_JPEG_LS:
+        return PROFILE_MJPEG_JPEG_LS;
+      default:
+        return PROFILE_UNKNOWN;
+    }
+  } else {
+    return PROFILE_UNKNOWN;
+  }
+}
+
+static int get_stride(int width, int align) {
+  return (width + align - 1) & (~(align - 1));
+}
+
+static int stcodec_send_data_to_decoder(AVCodecContext* avctx, uint8_t* buffer,
+                                        int size, int64_t pts, int64_t dts,
+                                        int64_t duration) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  STCODECDecoder* decoder = (STCODECDecoder*)st_context->decoder_ref->data;
+  int ret = 0;
+
+  if (!buffer && 0 == size) {
+    // a NULL packet, need to send EOS to decoder.
+    PACKET_SetEos(decoder->pPacket, 1);
+    PACKET_SetLength(decoder->pPacket, 0);
+  } else {
+    // debug
+    av_log(avctx, AV_LOG_DEBUG, "%x %x %x %x %x length = %d\n", *buffer,
+           *(buffer + 1), *(buffer + 2), *(buffer + 3), *(buffer + 4), size);
+
+    PACKET_SetDataPointer(decoder->pPacket, buffer);
+    PACKET_SetLength(decoder->pPacket, size);
+    if (pts != AV_NOPTS_VALUE) {
+      PACKET_SetPts(decoder->pPacket, pts);
+    } else if (dts != AV_NOPTS_VALUE) {
+      PACKET_SetPts(decoder->pPacket, dts);
+      decoder->use_dts = 1;
+    } else {
+      decoder->no_pts = 1;
+    }
+    PACKET_SetEos(decoder->pPacket, 0);
+    av_log(avctx, AV_LOG_DEBUG, "input pts : %ld\n", pts);
+  }
+
+  // send packet to decoder, the packet will be copyed in decoder, so can
+  // release it here.
+  ret = VDEC_Decode(decoder->pVdecCtx, PACKET_GetBaseData(decoder->pPacket));
+  if (ret) {
+    if (ret == MPP_DATAQUEUE_FULL) {
+      // input unblock mode will go here.
+      av_log(avctx, AV_LOG_DEBUG, "Buffer full writing %d bytes to decoder\n",
+             size);
+      ret = AVERROR(EAGAIN);  // here something wrong, maybe return 0
+    } else {
+      ret = AVERROR_UNKNOWN;
+    }
+  } else {
+    av_log(avctx, AV_LOG_DEBUG, "Write %d bytes to decoder\n", size);
+  }
+
+  return ret;  // here something wrong, maybe return 0
+}
+
+static av_cold int stcodec_close_decoder(AVCodecContext* avctx) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  av_log(NULL, AV_LOG_DEBUG, "stcodec close decoder\n");
+  av_buffer_unref(&st_context->decoder_ref);
+  return 0;
+}
+
+static void stcodec_release_stcodec_decoder(void* opaque, uint8_t* data) {
+  STCODECDecoder* decoder = (STCODECDecoder*)data;
+
+  if (decoder->pPacket) {
+    // PACKET_Free(decoder->pPacket);
+    PACKET_Destory(decoder->pPacket);
+    decoder->pPacket = NULL;
+  }
+
+  if (decoder->pFrame) {
+    FRAME_Destory(decoder->pFrame);
+    decoder->pFrame = NULL;
+  }
+
+  if (decoder->pVdecCtx) {
+    // VDEC_ResetChannel(decoder->pVdecCtx);
+    VDEC_DestoryChannel(decoder->pVdecCtx);
+    decoder->pVdecCtx = NULL;
+  }
+
+  av_buffer_unref(&decoder->frames_ref);
+  av_buffer_unref(&decoder->device_ref);
+  av_free(decoder);
+}
+
+static av_cold int stcodec_init_decoder(AVCodecContext* avctx) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  STCODECDecoder* decoder = NULL;
+  MppCodingType codectype = CODING_UNKNOWN;
+  int ret;
+
+  if ((avctx->width > 4096 || avctx->height > 4096 || avctx->width < 640 ||
+       avctx->height < 320) &&
+      (avctx->width != 0 || avctx->height != 0)) {
+    av_log(avctx, AV_LOG_ERROR,
+           "STCODEC Decoder do not support the size (%d x %d), too big or too "
+           "small!\n",
+           avctx->width, avctx->height);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  avctx->pix_fmt = ff_get_format(avctx, ffcodec(avctx->codec)->pix_fmts);
+  av_log(avctx, AV_LOG_ERROR, "Use pixel format %d\n", avctx->pix_fmt);
+
+  // create a decoder and a ref to it
+  decoder = av_mallocz(sizeof(STCODECDecoder));
+  if (!decoder) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to av_mallocz STCODECDecoder\n");
+    ret = AVERROR(ENOMEM);
+    goto fail;
+  }
+
+  decoder->ePixFmt = avctx->pix_fmt;
+
+  st_context->decoder_ref = av_buffer_create(
+      (uint8_t*)decoder, sizeof(*decoder), stcodec_release_stcodec_decoder,
+      NULL, AV_BUFFER_FLAG_READONLY);
+  if (!st_context->decoder_ref) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to create ref of STCODECDecoder!\n");
+    av_free(decoder);
+    ret = AVERROR(ENOMEM);
+    goto fail;
+  }
+
+  av_log(avctx, AV_LOG_DEBUG, "Initializing STCODEC decoder.\n");
+
+  codectype = stcodec_get_codingtype(avctx);
+  if (CODING_UNKNOWN == codectype) {
+    av_log(avctx, AV_LOG_ERROR, "Unknown codec type (%d).\n", avctx->codec_id);
+    av_free(decoder);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  // Create the MPP context
+  decoder->pVdecCtx = VDEC_CreateChannel();
+  if (!decoder->pVdecCtx) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to create STCODEC VDEC channel.\n");
+    av_free(decoder);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  // set para
+  decoder->pVdecCtx->stVdecPara.eCodingType = codectype;
+  decoder->pVdecCtx->stVdecPara.nProfile = stcodec_get_profiletype(avctx);
+  decoder->pVdecCtx->stVdecPara.bInputBlockModeEnable = MPP_FALSE;
+  decoder->pVdecCtx->stVdecPara.bOutputBlockModeEnable = MPP_TRUE;
+  decoder->pVdecCtx->stVdecPara.nWidth = avctx->width;
+  decoder->pVdecCtx->stVdecPara.nHeight = avctx->height;
+  decoder->pVdecCtx->stVdecPara.nStride = get_stride(avctx->width, 8);
+  decoder->pVdecCtx->stVdecPara.eOutputPixelFormat =
+      stcodec_get_pixelformat(avctx);
+  if (avctx->pix_fmt == AV_PIX_FMT_DRM_PRIME) {
+    decoder->pVdecCtx->stVdecPara.eFrameBufferType =
+        MPP_FRAME_BUFFERTYPE_DMABUF_INTERNAL;
+  } else {
+    decoder->pVdecCtx->stVdecPara.eFrameBufferType =
+        MPP_FRAME_BUFFERTYPE_NORMAL_INTERNAL;
+  }
+  decoder->pVdecCtx->eCodecType = CODEC_V4L2_LINLONV5V7;
+  if (avctx->width >= 3840 || avctx->height >= 2160) {
+    av_log(avctx, AV_LOG_ERROR, "4K video, downscale!\n");
+    decoder->pVdecCtx->stVdecPara.nScale = 2;
+  } else {
+    decoder->pVdecCtx->stVdecPara.nScale = 1;
+  }
+  decoder->pVdecCtx->stVdecPara.nHorizonScaleDownRatio = 1;
+  decoder->pVdecCtx->stVdecPara.nVerticalScaleDownRatio = 1;
+  decoder->pVdecCtx->stVdecPara.nRotateDegree = 0;
+  decoder->pVdecCtx->stVdecPara.bThumbnailMode = 0;
+  decoder->pVdecCtx->stVdecPara.bIsInterlaced = MPP_FALSE;
+
+  // vdec init
+  ret = VDEC_Init(decoder->pVdecCtx);
+  if (ret) {
+    av_log(avctx, AV_LOG_ERROR,
+           "Failed to initialize STCODEC VDEV (ret = %d).\n", ret);
+    // VDEC_DestoryChannel(decoder->pVdecCtx);
+    // av_free(decoder);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  // mpp packet init
+  decoder->pPacket = PACKET_Create();
+  if (!decoder->pPacket) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to initialize STCODEC packet.\n");
+    // VDEC_DestoryChannel(decoder->pVdecCtx);
+    // av_free(decoder);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  // mpp frame init
+  decoder->pFrame = FRAME_Create();
+  if (!decoder->pFrame) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to initialize STCODEC frame.\n");
+    PACKET_Destory(decoder->pPacket);
+    // VDEC_DestoryChannel(decoder->pVdecCtx);
+    // av_free(decoder);
+    ret = AVERROR_UNKNOWN;
+    goto fail;
+  }
+
+  decoder->first_packet = 1;
+  decoder->duration = 0;
+  decoder->use_dts = 0;
+  decoder->flushed = 0;
+
+  av_log(avctx, AV_LOG_DEBUG, "STCODEC decoder initialized successfully.\n");
+
+  decoder->device_ref = av_hwdevice_ctx_alloc(AV_HWDEVICE_TYPE_DRM);
+  if (!decoder->device_ref) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to av_hwdevice_ctx_alloc\n");
+    ret = AVERROR(ENOMEM);
+    goto fail;
+  }
+  ret = av_hwdevice_ctx_init(decoder->device_ref);
+  if (ret < 0) {
+    av_log(avctx, AV_LOG_ERROR, "Failed to av_hwdevice_ctx_init\n");
+    goto fail;
+  }
+  /*
+    decoder->frames_ref = av_hwframe_ctx_alloc(decoder->device_ref);
+    if (!decoder->frames_ref) {
+      av_log(avctx, AV_LOG_ERROR, "Failed to av_hwframe_ctx_alloc\n");
+      ret = AVERROR(ENOMEM);
+      goto fail;
+    }
+
+    AVHWFramesContext* hwframes;
+    hwframes = (AVHWFramesContext*)decoder->frames_ref->data;
+    hwframes->format = AV_PIX_FMT_DRM_PRIME;
+    hwframes->sw_format = AV_PIX_FMT_NV12;
+    hwframes->width = avctx->width;
+    hwframes->height = avctx->height;
+    ret = av_hwframe_ctx_init(decoder->frames_ref);
+    if (ret < 0) goto fail;
+  */
+  av_log(avctx, AV_LOG_DEBUG, "Initialized successfully.\n");
+  return 0;
+
+fail:
+  av_log(avctx, AV_LOG_ERROR,
+         "Failed to initialize STCODEC decoder, please check!\n");
+  return ret;
+}
+
+static int stcodec_send_packet(AVCodecContext* avctx, const AVPacket* avpkt) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  STCODECDecoder* decoder = (STCODECDecoder*)st_context->decoder_ref->data;
+  int ret;
+  av_log(avctx, AV_LOG_DEBUG,
+         "start send packet, pts(%ld) dts(%ld) duration(%ld)\n", avpkt->pts,
+         avpkt->dts, avpkt->duration);
+
+  // handle EOF
+  if (!avpkt->size) {
+    av_log(avctx, AV_LOG_ERROR, "Get EOS from parser!\n");
+    decoder->eos_reached = 1;
+    // write a NULL data to decoder to inform it the EOS.
+    ret = stcodec_send_data_to_decoder(avctx, NULL, 0, 0, 0, 0);
+    if (ret)
+      av_log(avctx, AV_LOG_ERROR, "Failed to send EOS to decoder (ret = %d)\n",
+             ret);
+    return ret;
+  }
+
+  // on first packet, send extradata
+  if (decoder->first_packet) {
+    if (avctx->extradata_size) {
+      if (avctx->codec_id == AV_CODEC_ID_H264 ||
+          avctx->codec_id == AV_CODEC_ID_HEVC ||
+          avctx->codec_id == AV_CODEC_ID_MPEG4 ||
+          avctx->codec_id == AV_CODEC_ID_VC1) {
+        ret = stcodec_send_data_to_decoder(avctx, avctx->extradata,
+                                           avctx->extradata_size, avpkt->pts,
+                                           avpkt->dts, avpkt->duration);
+        if (ret) {
+          av_log(avctx, AV_LOG_ERROR,
+                 "Failed to write extradata to decoder (ret = %d)\n", ret);
+          return ret;
+        } else {
+          av_log(avctx, AV_LOG_ERROR,
+                 "OK! Write extradata to decoder (size = %d)\n",
+                 avctx->extradata_size);
+        }
+      }
+    }
+    decoder->first_packet = 0;
+  }
+
+  // now send packet
+  ret = stcodec_send_data_to_decoder(avctx, avpkt->data, avpkt->size,
+                                     avpkt->pts, avpkt->dts, avpkt->duration);
+  if (ret) {
+    av_log(avctx, AV_LOG_DEBUG,
+           "Failed to write data to decoder (code = %d (%s))\n", ret,
+           av_err2str(ret));
+  } else {
+    av_log(avctx, AV_LOG_DEBUG,
+           "OK! Write data to decoder, (size = %d) (ret = %d)\n", avpkt->size,
+           ret);
+  }
+
+  return ret;
+}
+
+static void stcodec_release_frame(void* opaque, uint8_t* data) {
+  AVDRMFrameDescriptor* desc = (AVDRMFrameDescriptor*)data;
+  AVBufferRef* framecontextref = (AVBufferRef*)opaque;
+  STCODECFrameContext* framecontext =
+      (STCODECFrameContext*)framecontextref->data;
+
+  av_log(NULL, AV_LOG_DEBUG, "stcodec release frame\n");
+
+  VDEC_ReturnOutputFrame(
+      ((STCODECDecoder*)framecontext->decoder_ref->data)->pVdecCtx,
+      FRAME_GetBaseData(framecontext->pFrame));
+
+  if (((STCODECDecoder*)framecontext->decoder_ref->data)->ePixFmt ==
+      AV_PIX_FMT_DRM_PRIME) {
+    av_free(desc);
+  }
+  FRAME_Destory(framecontext->pFrame);
+  av_buffer_unref(&framecontext->decoder_ref);
+  av_buffer_unref(&framecontextref);
+}
+
+static int stcodec_receive_frame(AVCodecContext* avctx, AVFrame* frame) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  STCODECDecoder* decoder = (STCODECDecoder*)st_context->decoder_ref->data;
+  AVPacket pkt = {0};
+  int freeslots;
+  STCODECFrameContext* framecontext = NULL;
+  AVBufferRef* framecontextref = NULL;
+  MppFrame* mppframe = FRAME_Create();
+  AVDRMFrameDescriptor* desc = NULL;
+  AVDRMLayerDescriptor* layer = NULL;
+  int ret = -1;
+
+  av_log(avctx, AV_LOG_DEBUG, "start receive frame\n");
+
+  if (!decoder->eos_reached) {
+    // we get the available input queue num in decoder
+    VDEC_GetParam(decoder->pVdecCtx, &(decoder->pVdecPara));
+    freeslots = decoder->pVdecPara->nInputQueueLeftNum;
+    av_log(avctx, AV_LOG_DEBUG, "Input queue left %d seat!!\n", freeslots);
+
+    if (freeslots > 0) {
+      ret = ff_decode_get_packet(avctx, &pkt);
+      if (ret < 0 && ret != AVERROR_EOF) {
+        av_log(avctx, AV_LOG_DEBUG,
+               "Failed to ff_decode_get_packet (ret = %d (%s))\n", ret,
+               av_err2str(ret));
+        goto fail;
+      }
+
+      ret = stcodec_send_packet(avctx, &pkt);
+      av_packet_unref(&pkt);
+
+      if (ret < 0) {
+        av_log(avctx, AV_LOG_ERROR,
+               "Failed to send packet to decoder (code = %d (%s))\n", ret,
+               av_err2str(ret));
+        goto fail;
+      }
+    }
+  }
+
+  do {
+    ret =
+        VDEC_RequestOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+    if (ret == MPP_CODER_NULL_DATA) {
+      av_log(avctx, AV_LOG_ERROR, "null data, return\n");
+      VDEC_ReturnOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+    }
+  } while (ret == MPP_CODER_NO_DATA || ret == MPP_CODER_NULL_DATA);
+  av_log(avctx, AV_LOG_DEBUG, "Request frame (ret = %d)\n", ret);
+
+  if (ret == MPP_CODER_EOS && mppframe == NULL) {
+    av_log(avctx, AV_LOG_ERROR, "EOS 1!\n");
+    ret = AVERROR_EOF;
+    goto fail;
+  }
+
+  if (ret == MPP_CODER_EOS ||
+      FRAME_GetEos(mppframe) == 1 /* || decoder->eos_reached*/) {
+    av_log(avctx, AV_LOG_ERROR, "EOS 2!\n");
+    VDEC_ReturnOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+    ret = AVERROR_EOF;
+    goto fail;
+  }
+
+  if (ret == MPP_ERROR_FRAME) {
+    av_log(avctx, AV_LOG_ERROR, "error frame, need discard\n");
+    VDEC_ReturnOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+    ret = AVERROR(EAGAIN);
+    goto fail;
+  }
+
+  if (ret == MPP_RESOLUTION_CHANGED) {
+    av_log(avctx, AV_LOG_ERROR,
+           "resolution changed!!!!!!!!!!!!!!!!!!!!!!!!!!!\n");
+    AVHWFramesContext* hwframes;
+
+    av_log(avctx, AV_LOG_DEBUG, "Decoder noticed an info change (%dx%d)\n",
+           decoder->pVdecCtx->stVdecPara.nWidth,
+           decoder->pVdecCtx->stVdecPara.nHeight);
+
+    avctx->width = decoder->pVdecCtx->stVdecPara.nWidth;
+    avctx->height = decoder->pVdecCtx->stVdecPara.nHeight;
+
+    av_buffer_unref(&decoder->frames_ref);
+
+    decoder->frames_ref = av_hwframe_ctx_alloc(decoder->device_ref);
+    if (!decoder->frames_ref) {
+      av_log(avctx, AV_LOG_ERROR,
+             "av_hwframe_ctx_alloc failed, please check\n");
+      ret = AVERROR(ENOMEM);
+      goto fail;
+    }
+
+    hwframes = (AVHWFramesContext*)decoder->frames_ref->data;
+    hwframes->format = AV_PIX_FMT_DRM_PRIME;
+    hwframes->sw_format = AV_PIX_FMT_NV12;
+    hwframes->width = avctx->width;
+    hwframes->height = avctx->height;
+    ret = av_hwframe_ctx_init(decoder->frames_ref);
+    if (ret < 0) {
+      av_log(avctx, AV_LOG_ERROR, "av_hwframe_ctx_init failed, please check\n");
+      goto fail;
+    }
+
+    // here decoder is fully initialized, we need to feed it again with data
+    ret = AVERROR(EAGAIN);
+    goto fail;
+  }
+
+  if (!ret) {
+    av_log(NULL, AV_LOG_DEBUG, "stcodec request a frame\n");
+    if (!decoder->flushed) {
+      // setup general frame fields
+      frame->format = avctx->pix_fmt;
+      frame->width = avctx->width;
+      frame->height = avctx->height;
+      if (!decoder->no_pts) {
+        if (!decoder->use_dts) {
+          frame->pts = FRAME_GetPts(mppframe);
+        } else {
+          frame->pkt_dts = FRAME_GetPts(mppframe);
+        }
+      } else {
+      }
+      av_log(avctx, AV_LOG_DEBUG, "frame pts:%ld pkt_dts:%ld\n", frame->pts,
+             frame->pkt_dts);
+
+      framecontextref = av_buffer_allocz(sizeof(*framecontext));
+      if (!framecontextref) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to alloc AvBufferRef\n");
+        ret = AVERROR(ENOMEM);
+        goto fail;
+      }
+
+      // MPP decoder needs to be closed only when all frames have been released.
+      framecontext = (STCODECFrameContext*)framecontextref->data;
+      framecontext->decoder_ref = av_buffer_ref(st_context->decoder_ref);
+      framecontext->pFrame = mppframe;
+
+      if (avctx->pix_fmt == AV_PIX_FMT_DRM_PRIME) {
+        desc = av_mallocz(sizeof(AVDRMFrameDescriptor));
+        if (!desc) {
+          ret = AVERROR(ENOMEM);
+          goto fail;
+        }
+
+        desc->nb_objects = 1;
+        desc->objects[0].fd = FRAME_GetFD(mppframe, 0);
+        desc->objects[0].size = frame->width * frame->height * 3 / 2;
+        av_log(avctx, AV_LOG_DEBUG, "fd=%d size=%ld\n", desc->objects[0].fd,
+               desc->objects[0].size);
+        desc->nb_layers = 1;
+        layer = &desc->layers[0];
+        layer->format = DRM_FORMAT_NV12;
+        layer->nb_planes = 2;
+
+        layer->planes[0].object_index = 0;
+        layer->planes[0].offset = 0;
+        layer->planes[0].pitch = frame->width;
+
+        layer->planes[1].object_index = 0;
+        layer->planes[1].offset = frame->width * frame->height;
+        layer->planes[1].pitch = frame->width;
+
+        frame->data[0] = (uint8_t*)desc;
+        frame->buf[0] = av_buffer_create((uint8_t*)desc, sizeof(*desc),
+                                         stcodec_release_frame, framecontextref,
+                                         AV_BUFFER_FLAG_READONLY);
+      } else if (avctx->pix_fmt == AV_PIX_FMT_NV12) {
+        frame->linesize[0] = get_stride(avctx->width, 8);
+        frame->linesize[1] = get_stride(avctx->width, 8);
+        frame->data[0] = FRAME_GetDataPointer(mppframe, 0);
+        frame->data[1] = FRAME_GetDataPointer(mppframe, 1);
+        frame->buf[0] = av_buffer_create(
+            (uint8_t*)(frame->data[0]), sizeof(frame->data[0]),
+            stcodec_release_frame, framecontextref, AV_BUFFER_FLAG_READONLY);
+      } else if (avctx->pix_fmt == AV_PIX_FMT_YUV420P) {
+        frame->linesize[0] = get_stride(avctx->width, 8);
+        frame->linesize[1] = get_stride(avctx->width, 8) / 2;
+        frame->linesize[2] = get_stride(avctx->width, 8) / 2;
+        frame->data[0] = FRAME_GetDataPointer(mppframe, 0);
+        frame->data[1] = FRAME_GetDataPointer(mppframe, 1);
+        frame->data[2] = FRAME_GetDataPointer(mppframe, 2);
+
+        frame->buf[0] = av_buffer_create(
+            (uint8_t*)(frame->data[0]), sizeof(frame->data[0]),
+            stcodec_release_frame, framecontextref, AV_BUFFER_FLAG_READONLY);
+      }
+
+      if (!frame->buf[0]) {
+        av_log(avctx, AV_LOG_ERROR,
+               "Failed to create AVDRMFrameDescriptor ref\n");
+        ret = AVERROR(ENOMEM);
+        goto fail;
+      }
+
+      frame->hw_frames_ctx = av_buffer_ref(decoder->frames_ref);
+      if (!frame->hw_frames_ctx) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to create hw_frames_ctx\n");
+        ret = AVERROR(ENOMEM);
+        goto fail;
+      }
+    } else if (decoder->flushed < DROPFRAME_NUM_AFTER_FLUSH) {
+      VDEC_ReturnOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+      av_log(avctx, AV_LOG_ERROR, "drop some frames after flush, num=%d\n",
+             decoder->flushed);
+      decoder->flushed++;
+      ret = AVERROR(EAGAIN);
+      goto fail;
+    } else if (decoder->flushed == DROPFRAME_NUM_AFTER_FLUSH) {
+      VDEC_ReturnOutputFrame(decoder->pVdecCtx, FRAME_GetBaseData(mppframe));
+      ret = AVERROR(EAGAIN);
+      av_log(avctx, AV_LOG_ERROR,
+             "drop some frames after flush finish, num=%d\n", decoder->flushed);
+      decoder->flushed = 0;
+      goto fail;
+    }
+  } else {
+    av_log(avctx, AV_LOG_ERROR, "should not be here, ret = %d\n", ret);
+    ret = AVERROR(EAGAIN);
+    goto fail;
+  }
+
+  return 0;  // return AVERROR(EAGAIN);
+
+fail:
+  if (mppframe) FRAME_Destory(mppframe);
+  if (framecontext && framecontext->pFrame) FRAME_Destory(framecontext->pFrame);
+  if (framecontext) av_buffer_unref(&framecontext->decoder_ref);
+  if (framecontextref) av_buffer_unref(&framecontextref);
+  if (desc) av_free(desc);
+
+  return ret;
+}
+
+static void stcodec_flush(AVCodecContext* avctx) {
+  STCODECDecodeContext* st_context = avctx->priv_data;
+  STCODECDecoder* decoder = (STCODECDecoder*)st_context->decoder_ref->data;
+  int ret = -1;
+
+  av_log(avctx, AV_LOG_ERROR, "Flush.\n");
+
+  ret = VDEC_Flush(decoder->pVdecCtx);
+  if (ret)
+    av_log(avctx, AV_LOG_ERROR, "Failed to flush VDEC Channel (code = %d)\n",
+           ret);
+
+  decoder->eos_reached = 0;
+  decoder->no_pts = 0;
+  decoder->first_packet = 1;
+  decoder->duration = 0;
+  decoder->use_dts = 0;
+  decoder->flushed = 1;
+}
+
+static const AVCodecHWConfigInternal* const stcodec_hw_configs[] = {
+    HW_CONFIG_INTERNAL(DRM_PRIME), NULL};
+
+#define STCODEC_DEC_CLASS(NAME)                       \
+  static const AVClass stcodec_##NAME##_dec_class = {  \
+      .class_name = "stcodec_" #NAME "_dec",           \
+      .version = LIBAVUTIL_VERSION_INT,               \
+  };
+
+#define STCODEC_DEC(NAME, ID, BSFS)                                           \
+  STCODEC_DEC_CLASS(NAME)                                                     \
+  const FFCodec ff_##NAME##_stcodec_decoder = {                               \
+      .p.name = #NAME "_stcodec",                                             \
+      CODEC_LONG_NAME(#NAME " (stcodec decoder)"),                            \
+      .p.type = AVMEDIA_TYPE_VIDEO,                                           \
+      .p.id = ID,                                                             \
+      .priv_data_size = sizeof(STCODECDecodeContext),                         \
+      .init = stcodec_init_decoder,                                           \
+      .close = stcodec_close_decoder,                                         \
+      FF_CODEC_RECEIVE_FRAME_CB(stcodec_receive_frame),                       \
+      .flush = stcodec_flush,                                                 \
+      .p.priv_class = &stcodec_##NAME##_dec_class,                            \
+      .p.capabilities = AV_CODEC_CAP_DELAY | AV_CODEC_CAP_AVOID_PROBING |     \
+                        AV_CODEC_CAP_HARDWARE,                                \
+      CODEC_PIXFMTS(AV_PIX_FMT_DRM_PRIME, AV_PIX_FMT_NV12,                    \
+                    AV_PIX_FMT_YUV420P),                                      \
+      .hw_configs = stcodec_hw_configs,                                       \
+      .bsfs = BSFS,                                                           \
+      .p.wrapper_name = "stcodec",                                            \
+      .caps_internal = FF_CODEC_CAP_INIT_CLEANUP |                             \
+                       FF_CODEC_CAP_NOT_INIT_THREADSAFE,                     \
+  };
+
+STCODEC_DEC(h264, AV_CODEC_ID_H264, "h264_mp4toannexb")
+STCODEC_DEC(hevc, AV_CODEC_ID_HEVC, "hevc_mp4toannexb")
+STCODEC_DEC(vp8, AV_CODEC_ID_VP8, NULL)
+STCODEC_DEC(vp9, AV_CODEC_ID_VP9, NULL)
+STCODEC_DEC(mjpeg, AV_CODEC_ID_MJPEG, NULL)
+STCODEC_DEC(mpeg4, AV_CODEC_ID_MPEG4, NULL)
